BlueShift-iOS-SDK/BlueShift.m

```diff
@@ -42,113 +58,122 @@ - (void)setAppDelegate {
-    
-    // setting the new delegate's old delegate with the original delegate we saved...
-    BlueShiftAppDelegate *blueShiftAppDelegate = (BlueShiftAppDelegate *)_newDelegate;
-    blueShiftAppDelegate.oldDelegate = oldDelegate;
-    
-    if (@available(iOS 10.0, *)) {
-        if(config.userNotificationDelegate) {
-            blueShiftAppDelegate.userNotificationDelegate = config.userNotificationDelegate;
-        } else {
-            blueShiftAppDelegate.userNotificationDelegate = blueShiftUserNotificationCenterDelegate;
+    @try {
+        // Validate the API key
+        if ([config validateConfigDetails] == NO) {
+            return ;
         }
-    }
-    
-    if(config.blueShiftPushDelegate) {
-        blueShiftAppDelegate.blueShiftPushDelegate = config.blueShiftPushDelegate;
-    }
-    if (config.blueshiftUniversalLinksDelegate) {
-        blueShiftAppDelegate.blueshiftUniversalLinksDelegate = config.blueshiftUniversalLinksDelegate;
-    }
-    if (config.blueshiftDeviceIdSource) {
-        [[BlueShiftDeviceData currentDeviceData] setBlueshiftDeviceIdSource:config.blueshiftDeviceIdSource];
         
-        //Custom device id provision for DeviceIDSourceCUSTOM
-        if (config.blueshiftDeviceIdSource == BlueshiftDeviceIdSourceCustom) {
-            if (config.customDeviceId && ![config.customDeviceId isEqualToString:@""]) {
-                [[BlueShiftDeviceData currentDeviceData] setCustomDeviceID:config.customDeviceId];
-                [BlueshiftLog logInfo: [NSString stringWithFormat:@"CUSTOM device id is set as - %@",config.customDeviceId] withDetails:nil methodName:nil];
+        // Set config
+        _sharedBlueShiftInstance.config = config;
+        
+        // Set up device id
+        if (config.blueshiftDeviceIdSource) {
+            [[BlueShiftDeviceData currentDeviceData] setBlueshiftDeviceIdSource:config.blueshiftDeviceIdSource];
+            
+            //Custom device id provision for DeviceIDSourceCUSTOM
+            if (config.blueshiftDeviceIdSource == BlueshiftDeviceIdSourceCustom) {
+                if (config.customDeviceId && ![config.customDeviceId isEqualToString:@""]) {
+                    [[BlueShiftDeviceData currentDeviceData] setCustomDeviceID:config.customDeviceId];
+                    [BlueshiftLog logInfo: [NSString stringWithFormat:@"CUSTOM device id is set as - %@",config.customDeviceId] withDetails:nil methodName:nil];
+                } else {
+                    [BlueshiftLog logError:nil withDescription:@"ERROR: CUSTOM device id is not provided" methodName:nil];
+                }
+            }
+        }
+        
+        // Set BlueshiftAppDelegate
+        _sharedBlueShiftInstance.appDelegate = [[BlueShiftAppDelegate alloc] init];
+        _sharedBlueShiftInstance.appDelegate.mainAppDelegate = [UIApplication sharedApplication].delegate;
+        
+        // Initialise core data
+        [_sharedBlueShiftInstance.appDelegate initializeCoreData];
```

Comment:
this may involve disc IO operations, right? will it cause any issues when called from the main thread?

Reply:
Initializing core data on the main thread should not cause any issue. It is just creating the core data related objects based on sqlite db.